### ISSUE TITLE
page.driver.debug not working

### ISSUE THREAD

**Issue by mrbrdo** [oesmith/puffing-billy#43]
I'm using this with Poltergeist and Capybara, and the proxy seems to be working. However, first, after I make a request, I cannot do `Billy.proxy.port` anymore, I get:

`RuntimeError: eventmachine not initialized: evma_get_sockname from /x/gems/puffing-billy-0.2.3/lib/billy/proxy.rb:32:in get_sockname`

Then, when I use `page.driver.debug`, it will open my browser as normal but it is then not able to connect to the server and immediately shows the could not connect page.

It works fine if I don't set the proxy for Poltergeist.

**Comment by ronwsmith** [comment]
@mrbrdo, please see if this issue still occurs on v0.6.1.  If so, provide your Billy.config block.  Thanks!

**Comment by ronwsmith** [comment]
Closing -- stale issue.

**Comment by adamlangsner** [comment]
I think I'm having a similar problem. I'm using puffing-billy 0.9.1
Here's some background that's hopefully helpful:
We're using capybara/poltergeist in production to build screen scrapers for javascript heavy websites. We use puffing billy in our tests to cache the sites we're scraping to make our tests run quicker and allow them to run without an internet connection.
Everything worked with one scraper but after building a second scraper I started getting intermittent test failures (each scraper was scraping a different website). Since we randomize our tests, it depended on the run order of the tests. When the tests did fail, the first scraper test passed and then the second one failed on this line: https://github.com/oesmith/puffing-billy/blob/master/lib/billy/browsers/capybara.rb#L30 when it's trying to get the port. I got the  `eventmachine not initialized: evma_get_sockname` error message. Also, when I would run each test independently they both worked.

I had registered this driver in an initializer:
```rb
Capybara.register_driver :poltergeist_scraper do |app|
  Capybara::Poltergeist::Driver.new(app, js_errors: false)
end
```

and in my scrapers I had a method like this:
```rb
def session
  @session ||= Capybara::Session.new(:poltergeist_scraper)
end
```

then in my tests I'd create a capybara session with the puffing billy driver:
```rb
let(:billy_session) do
  Capybara::Session.new(:poltergeist_billy).tap do |session|
    session.driver.options[:js_errors] = false
  end
end
```

and I'd make the scraper use that instead of the `poltergeist_scraper`:
```rb
before do
  allow(Capybara::Session).to receive(:new).and_return(billy_session)
end
```

I'm not too familiar with event machine. One naive thing I tried was commenting out the `#reset` call that happens after each test run: https://github.com/oesmith/puffing-billy/blob/master/lib/billy/init/rspec.rb#L13 but that didn't work.

Not really sure if I'm doing something wrong or if there's a bug. Let me know if there's any other information I can provide to help with this.
Thanks in advance!

**Comment by ronwsmith** [comment]
@adamlangsner your setup seems like overkill.

You can just set up billy in your initializer and then skip the later two code sections you mentioned.

To do that, pass in the options block from https://github.com/oesmith/puffing-billy/blob/master/lib/billy/browsers/capybara.rb#L27-L32 with your `js_errors` flag.

**Comment by adamlangsner** [comment]
@ronwsmith Thanks! I've simplified my configuration based on your suggesting. I now just have 2 things:
in `config/initializers/capybara.rb` I set up the driver for production use:
```rb
Capybara.register_driver :poltergeist_scraper do |app|
  Capybara::Poltergeist::Driver.new(app, js_errors: false)
end
```

in my `spec/rails_helper.b` I overwrite it adding in the proxy:
```rb
Capybara.register_driver :poltergeist_scraper do |app|
  Capybara::Poltergeist::Driver.new(app,
    js_errors: false,
    phantomjs_options: [
      "--ignore-ssl-errors=yes",
      "--proxy=#{Billy.proxy.host}:#{Billy.proxy.port}"
    ]
  )
end
```

And I removed the later 2 pieces of code in my specs as you suggested.
It works, but doesn't fix my initial issue. I'm still getting the `eventmachine not initialized: evma_get_sockname` error when I run my entire test suite

**Comment by ronwsmith** [comment]
Does it work fine when you comment out the `--proxy` line?  What information do you see in the logs from puffing-billy around the time the issue occurs?

**Comment by adamlangsner** [comment]
I removed the proxy and everything worked. I added logging and everything started working all of a sudden. I did notice a new pattern though. Sometimes when I ran tests I'd get new cache files because of dynamic parameters in the requests, I noticed that if I kept deleting these new cache files after each run I'd get the error each time, but if I kept them everything would pass on the next test run. I had been deleting these new cache files after each test because I was procrastinating on configuring the appropriate values for `ignore_params` and `merge_cached_responses_whitelist` and didn't want extra cruft in my cache dir. But I just now decided to give up on that and commit the files. Everything seems to be working now. Sorry for the trouble, but thanks for taking the time to respond and thanks for your tip on configuring the capybara drivers for tests, that really cleaned up my tests!
